File: android/uhabits-android/src/main/java/org/isoron/uhabits/widgets/views/CurrentStreakWidgetView.java
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.widgets.views;

import android.content.Context;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.isoron.androidbase.utils.ColorUtils;
import org.isoron.androidbase.utils.StyledResources;
import org.isoron.uhabits.R;
import org.isoron.uhabits.activities.common.views.RingView;
import org.isoron.uhabits.core.models.Checkmark;

import static org.isoron.androidbase.utils.InterfaceUtils.getDimension;

public class CurrentStreakWidgetView extends HabitWidgetView
{
    private float percentage;

    @Nullable
    private String name;

    private RingView ring;

    private TextView label;

    private int checkmarkValue;
    private int startColor = 0x388e3c + (255 << 24); // add full alpha
    private int endColor = 0xc62828 + (255 << 24); // add full alpha
    private String currentStreak;
    private float timeoutPercentage;

    public CurrentStreakWidgetView(Context context)
    {
        super(context);
        init();
    }

    public CurrentStreakWidgetView(Context context, AttributeSet attrs)
    {
        super(context, attrs);
        init();
    }

    public void refresh()
    {
        if (backgroundPaint == null || frame == null || ring == null) return;

        StyledResources res = new StyledResources(getContext());

        int bgColor;
        int fgColor;

        bgColor = ColorUtils.interPolateHSV(startColor, endColor, timeoutPercentage);
        switch (checkmarkValue)
        {
            case Checkmark.CHECKED_EXPLICITLY:
                fgColor = res.getColor(R.attr.highContrastReverseTextColor);
                setShadowAlpha(0x4f);
                break;

            case Checkmark.CHECKED_IMPLICITLY:
            case Checkmark.UNCHECKED:
            default:
                fgColor = res.getColor(R.attr.mediumContrastTextColor);
                break;
        }

        backgroundPaint.setColor(bgColor);
        frame.setBackgroundDrawable(background);
        ring.setPercentage(percentage);
        ring.setColor(fgColor);
        ring.setBackgroundColor(bgColor);
        ring.setText(currentStreak);

        label.setText(name);
        label.setTextColor(fgColor);

        requestLayout();
        postInvalidate();
    }

    public void setCurrentStreak(String currentStreak)
    {
        this.currentStreak = currentStreak;
    }

    public void setCheckmarkValue(int checkmarkValue)
    {
        this.checkmarkValue = checkmarkValue;
    }

    public void setName(@NonNull String name)
    {
        this.name = name;
    }

    public void setPercentage(float percentage)
    {
        this.percentage = percentage;
    }
    public void setTimeOutPercentage(float percentage)
    {
        this.timeoutPercentage = percentage;
    }

    @Override
    @NonNull
    protected Integer getInnerLayoutId()
    {
        return R.layout.widget_checkmark;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec)
    {
        int width = MeasureSpec.getSize(widthMeasureSpec);
        int height = MeasureSpec.getSize(heightMeasureSpec);

        float w = width;
        float h = width * 1.25f;
        float scale = Math.min(width / w, height / h);

        w *= scale;
        h *= scale;

        if (h < getDimension(getContext(), R.dimen.checkmarkWidget_heightBreakpoint))
            ring.setVisibility(GONE);
        else
            ring.setVisibility(VISIBLE);

        widthMeasureSpec =
            MeasureSpec.makeMeasureSpec((int) w, MeasureSpec.EXACTLY);
        heightMeasureSpec =
            MeasureSpec.makeMeasureSpec((int) h, MeasureSpec.EXACTLY);

        float textSize = 0.15f * h;
        float maxTextSize = getDimension(getContext(), R.dimen.smallerTextSize);
        textSize = Math.min(textSize, maxTextSize);

        label.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize);
        ring.setTextSize(textSize);
        ring.setThickness(0.15f * textSize);

        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    private void init()
    {
        ring = (RingView) findViewById(R.id.scoreRing);
        label = (TextView) findViewById(R.id.label);

        if (ring != null) ring.setIsTransparencyEnabled(true);

        if (isInEditMode())
        {
            percentage = 0.75f;
            name = "Wake up early";
            checkmarkValue = Checkmark.CHECKED_EXPLICITLY;
            refresh();
        }
    }
}

File: uhabits-android/src/main/java/org/isoron/uhabits/widgets/views/CurrentStreakWidgetView.kt
/*
 * Copyright (C) 2016-2025 Álinson Santos Xavier <git@axavier.org>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.isoron.uhabits.widgets.views

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.widget.TextView
import org.isoron.uhabits.R
import org.isoron.uhabits.utils.InterfaceUtils
import org.isoron.uhabits.utils.StyledResources
import org.isoron.uhabits.core.models.PaletteColor
import org.isoron.uhabits.utils.toFixedAndroidColor
import kotlin.math.min
import kotlin.math.roundToInt

class CurrentStreakWidgetView : HabitWidgetView {
    var activeColor: Int = 0
    var currentStreak = 0
    var name: String? = null
    var isCompleted = false

    private lateinit var currentStreakText: TextView
    private lateinit var streakBgIcon: TextView
    private lateinit var label: TextView

    constructor(context: Context?) : super(context) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    fun refresh() {
        if (backgroundPaint == null || frame == null) return
        val res = StyledResources(context)
        val bgColor: Int
        val fgColor: Int
        setShadowAlpha(0x4f)

        if (isCompleted) {
            bgColor = activeColor
            fgColor = res.getColor(R.attr.contrast0)
            backgroundPaint!!.color = bgColor
            frame!!.setBackgroundDrawable(background)
        } else {
            bgColor = res.getColor(R.attr.cardBgColor)
            fgColor = res.getColor(R.attr.contrast60)
        }

        streakBgIcon.typeface = InterfaceUtils.getFontAwesome(context)
        streakBgIcon.setTextColor(getFireIconColor())
        streakBgIcon.alpha = 0.6f // Increased alpha since we are using specific colors

        currentStreakText.text = currentStreak.toString()
        currentStreakText.setTextColor(fgColor)

        label.text = name
        label.setTextColor(fgColor)

        requestLayout()
        postInvalidate()
    }

    private fun getFireIconColor(): Int {
        val red = PaletteColor(0).toFixedAndroidColor()
        val deepOrange = PaletteColor(1).toFixedAndroidColor()
        val orange = PaletteColor(2).toFixedAndroidColor()
        val amber = PaletteColor(3).toFixedAndroidColor()
        val yellow = PaletteColor(4).toFixedAndroidColor()
        val pink = PaletteColor(15).toFixedAndroidColor()

        val isWarmColor = activeColor == red || activeColor == deepOrange || activeColor == orange || activeColor == amber
                || activeColor == yellow || activeColor == pink

        return if (isWarmColor) {
            if (isCompleted) {
                Color.parseColor("#0D47A1") // Dark Blue
            } else {
                Color.parseColor("#BBDEFB") // Light Blue
            }
        } else {
            if (isCompleted) {
                Color.parseColor("#FF6D00") // Bright Orange
            } else {
                Color.parseColor("#FFE0B2") // Light Orange
            }
        }
    }

    override val innerLayoutId: Int
        get() = R.layout.widget_current_streak

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var width = MeasureSpec.getSize(widthMeasureSpec)
        var height = MeasureSpec.getSize(heightMeasureSpec)
        if (height >= width) {
            height = min(height, (width * 1.5).roundToInt())
        } else {
            width = min(width, height)
        }
        val labelTextSize = min(0.175f * width, InterfaceUtils.getDimension(context, R.dimen.smallTextSize))
        label.setTextSize(TypedValue.COMPLEX_UNIT_PX, labelTextSize)

        val streakTextSize = 0.4f * width
        currentStreakText.setTextSize(TypedValue.COMPLEX_UNIT_PX, streakTextSize)
        streakBgIcon.setTextSize(TypedValue.COMPLEX_UNIT_PX, streakTextSize * 1.5f)

        super.onMeasure(
            MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY),
            MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY)
        )
    }

    private fun init() {
        currentStreakText = findViewById<View>(R.id.current_streak_text) as TextView
        streakBgIcon = findViewById<View>(R.id.streak_bg_icon) as TextView
        label = findViewById<View>(R.id.label) as TextView
    }
}
